plutus-contract/src/Language/Plutus/Contract/StateMachine.hs

```diff
@@ -144,16 +124,17 @@ runStep ::
     -- ^ The state machine
     -> input
     -- ^ The input to apply to the state machine
-    -> Contract schema e state
+    -> Contract schema e (Maybe state)
 runStep smc input = do
-    -- the transaction returned by 'mkStep' includes an output with the payments
-    -- to the script address, so we only need to deal with the 'vaOtherPayments'
-    -- field of the value allocation here.
-    (typedTx, newState, ValueAllocation{vaOtherPayments}) <- mkStep smc input
-    let tx = case typedTx of
-            (Typed.TypedTxSomeOuts tx') -> Tx.fromLedgerTx (Typed.toUntypedTx tx')
-    submitTxConfirmed (tx <> vaOtherPayments)
-    pure newState
+    typedConstraints <- mkStep smc input
+    case typedConstraints of
+        Left c -> do -- no continuing output (final state)
+            submitTxConfirmed (toUntypedLedgerConstraints c)
+            pure Nothing
+        Right c -> do -- has continuing output (nonfinal state)
+            submitTxConfirmed (toUntypedLedgerConstraints c)
+            let TypedOrUntypedTxOuts{totTypedTxOuts} = tcOutputs c
+            pure $ Just $ tyTxOutData $ hfHead totTypedTxOuts
```

Comment:
See, this is a problem that arises because we're hiding the state inside the constraints, we have to reconstruct it. And above, we could return the final state, but we can't because it's thrown away already.